aineko/extras/connectors/websocket_client.py

```diff
@@ -0,0 +1,216 @@
+    with value `SECRET_VALUE`, you can inject it into the url field by passing
+    `wss://example.com?secret={$SECRET_NAME}` as the url. The connector will
+    then replace `{$SECRET_NAME}` with `SECRET_VALUE` before connecting to the
+    WebSocket.
+
+    Example usage in pipeline.yml:
+    ```yaml title="pipeline.yml"
+    pipeline:
+      nodes:
+        WebSocketClient:
+          class: aineko.extras.WebSocketClient
+          outputs:
+            - test_websocket
+          node_params:
+            url: "wss://example.com"
+            header:
+              auth: "Bearer {$SECRET_NAME}"
+            init_messages:
+                - {"Greeting": "Hello, world!"}
+    ```
+    """
+
+    retry_count = 0
+
+    def _pre_loop_hook(self, params: dict | None = None) -> None:
+        """Initialize the WebSocket connection.
+
+        Raises:
+            ValueError: If the params are invalid.
+        """
+        # Cast params to ParamsWebSocketClient type
+        try:
+            if params is not None:
+                self.ws_params = ParamsWebSocketClient(**params)
+            else:
+                raise ValueError(
+                    "No params provided to WebSocket connector. "
+                    "Node requires at least a url param."
+                )
+        except Exception as err:  # pylint: disable=broad-except
+            # Cast pydantic validation error to ValueError
+            # so that it can be properly caught by the node
+            # Note: this is required because pydantic errors
+            # are not picklable
+            raise ValueError(
+                "Failed to cast params to ParamsWebSocketClient type. "
+                f"The following error occurred: {err}"
+            ) from err
+
+        # Create the websocket subscription
+        self.ws = websocket.WebSocket()
+        self.create_subscription()
+
+    def _execute(self, params: dict | None = None) -> None:
+        """Polls and gets data from the WebSocket.
+
+        Raises:
+            ValueError: If the retry count exceeds the max retries.
+        """
+        try:
+            # Poll the websocket
+            raw_message = self.ws.recv()
+        except Exception as err:  # pylint: disable=broad-except
+            # If the connection is closed, reconnect
+            self.log(
+                "Websocket connection closed. "
+                f"Reconnecting in {self.ws_params.retry_sleep} "
+                f"seconds... The following error occurred: {err}",
+                level="error",
+            )
+            time.sleep(self.ws_params.retry_sleep)
+            self.create_subscription()
+            return
+
+        try:
+            # Parse the message and emit to producers
+            message = json.loads(raw_message)
+            if self.ws_params.metadata is not None:
+                message = {
+                    "metadata": self.ws_params.metadata,
```

Comment:
Here, it looks like the same metadata would be added to every message from this connection. Would there be a case for adding different metadata per message type, or based on the added complexity do you think this would be better handled by a separate data processing node?

Reply:
This is intended to help add context to the raw websocket messages that make it easier to parse in downstream nodes. i think i understand what you are suggesting, but i cant think of a good way to do this type of complex metadata injection in a way that in generalized enough for this node.